### PR TITLE
hof: init at 0.6.9-beta.1

### LINKED ISSUE #263593: Package request: hof
**Project description**

Framework that joins data models, schemas, code generation, and a task engine. Language and technology agnostic. 

**Metadata**

* homepage URL: https://docs.hofstadter.io/
* source URL: https://github.com/hofstadter-io/hof
* license: asl2
* platforms: linux, darwin

### PR DESCRIPTION
**Pull request by jfvillablanca**
## Description of changes
Resolves https://github.com/NixOS/nixpkgs/issues/263593

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/by-name/ho/hof/package.nix [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+  src = fetchFromGitHub {
+    owner  = "hofstadter-io";
+    repo   = "hof";
+    rev    = "v${version}";
+    hash   = "sha256-4yVP6DRHrsp52VxBhr7qppPhInYEsvPbIfxxQcRwHTw=";
+  };
+
+  vendorHash = "sha256-cDUcYwcxPn+9TEP5lhVJXofijCZX94Is+Qt41PqUgjI=";
+
+  subPackages = [ ".hof/shadow/cli/cmd/hof/main.go" ];
```
Comment (kirillrdy) [inline pkgs/by-name/ho/hof/package.nix:22]: should this be cmd/hof ?

https://github.com/hofstadter-io/hof/blob/_dev/.github/workflows/hof.yml#L69C25-L69C34

Comment (jfvillablanca) [reply pkgs/by-name/ho/hof/package.nix]: good catch, thanks!

**Comment by mbrydak** [comment]
hey @kirillrdy, don't want to be rude, just pinging you, cause it seems like everything is in order to merge this PR :)

**Review by kirillrdy — APPROVED** [review]

File: pkgs/by-name/ho/hof/package.nix [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+in
+buildGoModule {
+  inherit pname version;
+
+  src = fetchFromGitHub {
+    owner  = "hofstadter-io";
+    repo   = "hof";
+    rev    = "v${version}";
+    hash   = "sha256-4yVP6DRHrsp52VxBhr7qppPhInYEsvPbIfxxQcRwHTw=";
+  };
```
Comment (Janik-Haag) [inline pkgs/by-name/ho/hof/package.nix:18]:
```suggestion

buildGoModule rec {
  pname = "hof";
  version = "0.6.9-beta.1";

  src = fetchFromGitHub {
    owner  = "hofstadter-io";
    repo   = "hof";
    rev    = "v${version}";
    hash   = "sha256-4yVP6DRHrsp52VxBhr7qppPhInYEsvPbIfxxQcRwHTw=";
  };
```

you can use rec to make it recursive so you can reuse the attributes in the same attrSet. We usually don't use the let binding for pname or version.

**Comment by Janik-Haag** [comment]
Result of `nixpkgs-review pr 263683` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>hof</li>
  </ul>
</details>

File: pkgs/by-name/ho/hof/package.nix [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+  vendorHash = "sha256-cDUcYwcxPn+9TEP5lhVJXofijCZX94Is+Qt41PqUgjI=";
+
+  subPackages = [ "./cmd/hof/main.go" ];
+
+  postInstall = ''
+    mv $out/bin/main $out/bin/hof
+  '';
```
Comment (Janik-Haag) [inline pkgs/by-name/ho/hof/package.nix:26]:
```suggestion
  postInstall = ''
    mv $out/bin/main $out/bin/hof
    local INSTALL="$out/bin/hof"
    installShellCompletion --cmd hof \
      --bash <($out/bin/hof completion bash) \
      --fish <($out/bin/hof completion fish) \
      --zsh <($out/bin/hof completion zsh)
  '';
```
hove supports a bunch of completions so we should add them to the package.

File: pkgs/by-name/ho/hof/package.nix [hunk 4]
```diff
@@ -0,0 +1,35 @@
+{ lib
+, buildGoModule
+, fetchFromGitHub
```
Comment (Janik-Haag) [inline pkgs/by-name/ho/hof/package.nix:3]:
```suggestion
, fetchFromGitHub
, installShellFiles
```
needed for shell completions

File: pkgs/by-name/ho/hof/package.nix [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+buildGoModule {
+  inherit pname version;
+
+  src = fetchFromGitHub {
+    owner  = "hofstadter-io";
+    repo   = "hof";
+    rev    = "v${version}";
+    hash   = "sha256-4yVP6DRHrsp52VxBhr7qppPhInYEsvPbIfxxQcRwHTw=";
+  };
+
```
Comment (Janik-Haag) [inline pkgs/by-name/ho/hof/package.nix:19]:
```suggestion
  nativeBuildInputs = [ installShellFiles ];

```

**Review by Janik-Haag — CHANGES_REQUESTED** [review]

File: pkgs/by-name/ho/hof/package.nix (hunk 2 above)
Comment (kirillrdy) [reply pkgs/by-name/ho/hof/package.nix:18]: @Janik-Haag  Thank you for taking a look at this.
I was under impression that let is preferred over rec . Sorry to put @happysalada on the spot. https://github.com/NixOS/nixpkgs/pull/227858/

I personally don't mind let. Nor rec. I just hope we make it easier to understand for newcomers what is acceptable approach.

Comment (Janik-Haag) [reply pkgs/by-name/ho/hof/package.nix:18]: as far as I know rec is preferred in this case but I also don't care too much to be honest.

Comment (jfvillablanca) [reply pkgs/by-name/ho/hof/package.nix:18]: I usually have no preference between `let..in` vs `rec` and the expression is not too complex where `rec` would be hard to follow. I just subscribed to the "best practices" outlined [here](https://nix.dev/guides/best-practices#recursive-attribute-set-rec) but I'll defer to your suggestion 😀

Comment (happysalada) [reply pkgs/by-name/ho/hof/package.nix:18]: Haha, no worries. I personally prefer not using rec, but its definitely a matter of preference.

I would say if you feel strongly about it either way feel free to argue it, i think we are open in the community.

If you dont have any strong feelings about it, feel free to go with the flow, your PR will get merged faster :-)

Comment (happysalada) [reply pkgs/by-name/ho/hof/package.nix:18]: Just to try to bring to the debate.

Rec generates a bit of indirection on which variables is being used in different places, but it has the advantage of being more terse.
I think its a close call and really depends on what you prefer, terse vs obvious. (Obvious is also relative i think, its more obvious for new comers, but people used to it dont care that much anymore).

(The guy writing the guide you linked also thinks obvious beats terse)

**Comment by Janik-Haag** [comment]
Result of `nixpkgs-review pr 263683` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>hof</li>
  </ul>
</details>

**Review by Janik-Haag — APPROVED** [review]
